File: backend/src/ingestion/extractText.ts
/**
 * Port of `app.ingestion.extract_text`.
 */

function collapseWhitespace(text: string): string {
  return text.replace(/\s+/g, " ").trim();
}

function stripHtmlTags(html: string): string {
  return collapseWhitespace(
    html
      .replace(/<script[\s\S]*?<\/script>/gi, " ")
      .replace(/<style[\s\S]*?<\/style>/gi, " ")
      .replace(/<noscript[\s\S]*?<\/noscript>/gi, " ")
      .replace(/<[^>]+>/g, " "),
  );
}

export function extractTitleFromHtml(html: string): string {
  const match = html.match(/<title[^>]*>([\s\S]*?)<\/title>/i);
  if (!match?.[1]) return "";
  return collapseWhitespace(
    match[1].replace(/&nbsp;/gi, " ").replace(/&amp;/gi, "&"),
  );
}

/** Port of `extract_from_html`. */
export function extractFromHtml(html: string): string {
  if (!html.trim()) return "";

  const articleMatch = html.match(
    /<article[\s\S]*?<\/article>/i,
  );
  const mainMatch = html.match(/<main[\s\S]*?<\/main>/i);
  const chunk = articleMatch?.[0] ?? mainMatch?.[0] ?? html;

  return stripHtmlTags(chunk);
}

/** HTML-only; PDF/raw extraction runs in Python (`python/app/ingestion/extract_text.py`). */

File: python/app/ingestion/extract_text.py
"""Port of app.ingestion.extract_text — HTML, PDF, and raw text extraction."""

from __future__ import annotations

import html as html_module
import io
import re


def extract_from_html(html: str) -> str:
    if not (html or "").strip():
        return ""
    try:
        from trafilatura import extract

        text = extract(
            html,
            include_comments=False,
            include_tables=False,
            favor_precision=True,
        )
        if text:
            return html_module.unescape(text.strip())
    except Exception:
        pass

    # Fallback: strip tags
    cleaned = re.sub(r"<script[\s\S]*?</script>", " ", html, flags=re.I)
    cleaned = re.sub(r"<style[\s\S]*?</style>", " ", cleaned, flags=re.I)
    cleaned = re.sub(r"<[^>]+>", " ", cleaned)
    return re.sub(r"\s+", " ", cleaned).strip()


def extract_from_pdf(pdf_bytes: bytes) -> str:
    from pdfminer.high_level import extract_text

    text = extract_text(io.BytesIO(pdf_bytes)) or ""
    return extract_from_raw(text)


def extract_from_raw(text: str) -> str:
    if not text:
        return ""
    normalized = (
        text.replace("\r\n", "\n")
        .replace("&nbsp;", " ")
        .replace("&amp;", "&")
        .replace("&lt;", "<")
        .replace("&gt;", ">")
    )
    return re.sub(r"\s+", " ", normalized).strip()
